Added emergency stop support.

jim-rothrock opened the pull request:
I have added emergency stop support to the gazebo_ros_control and DefaultRobotHWSim plugins. gazebo_ros_control now resets the controllers when the robot's simulated hardware indicates that the emergency stop has been deactivated after being activated. DefaultRobotHWSim now subscribes to the Boolean topic specified by the eStopTopic SDF element. Messages received from the topic control the E-stop. When the E-stop is activated, the following occurs:
- If a joint's position is being controlled, DefaultRobotHWSim maintains the position the joint had when the E-stop was activated.
- If a joint's velocity is being controlled, the velocity is set to zero.
- If a joint's effort is being controlled, the effort is set to zero.

Commit:
Added emergency stop support.

adolfo-rt commented on gazebo_ros_control/include/gazebo_ros_control/robot_hw_sim.h:
```diff
@@ -34,10 +34,10 @@
-*/
+/// \brief Plugin template for hardware interfaces for ros_control and Gazebo
+
+/// \author Jonathon Brohen
```
Haha, now that we're at it, we could fix it to 'Jonathan Bohren'
@jbohren

jim-rothrock replied:
I'll fix it.

adolfo-rt commented on gazebo_ros_control/include/gazebo_ros_control/robot_hw_sim.h:
```diff
@@ -62,24 +62,59 @@ namespace gazebo_ros_control {
+    /// \return  \c true if the simulated robot hardware is initialized successfully, \c false if not.
     virtual bool initSim(
         const std::string& robot_namespace,
+        const sdf::ElementPtr sdf,
```
Instead of passing the SDF to `initSim`, how about simply passing the estop topic name?. It would be more consistent with how the `robot_namespace` parameter of this same method is obtained.

All plugin-specific fields are fetched from the SDF in [`GazeboRosControlPlugin::Load(...)`](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/indigo-devel/gazebo_ros_control/src/gazebo_ros_control_plugin.cpp#L58), and then stored and passed to the other relevant methods.

jim-rothrock replied:
Originally, I did that. In fact, the E-stop subscription and callback were in gazebo_ros_control_plugin.cpp. I changed the code in response to this comment in https://github.com/ros-simulation/gazebo_ros_pkgs/issues/258:

> Does your proposal only affect the DefaultRobotHWSim implementation?. I get the
> impression that it does, and I would agree with this course of action, and to not impose one
> E-stop implementation in the general case. This would mean that the eStopTopic SDF
> element would only be read by DefaultRobotHWSim implementation, and not by 
> RobotHWSim, correct?.

My original plan was to put as much of the E-stop code as possible into gazebo_ros_control_plugin.cpp, and only make default_robot_hardware_sim.cpp responsible for stopping joint motion. When I considered your comment, it occurred to me that alternative RobotHWSim plugins might want motion pause/resume controlled by some mechanism other than a topic. For example, initSim() could open a Qt window that contained _pause_ and _resume_ buttons. Here are the strategies I've considered:

1) gazebo_ros_control_plugin.cpp subscribes to the E-stop topic and contains the callback. It calls RobotHWSim's eStopActive(), passing it an argument telling it whether the E-stop is active. default_robot_hardware_sim.cpp is responsible to checking the state set by eStopActive(), pausing and resuming joint motion as necessary.

2) RobotHWSim is responsible for detecting E-stops, either by subscribing to a topic, or by some other means. It pauses or resumes joint motion, and informs gazebo_ros_control_plugin.cpp of the E-stop state via RobotHWSim's eStopActive() (which does not have a parameter in this case). gazebo_ros_control_plugin.cpp calls eStopActive() during each control loop iteration, and resets controllers in response.

I'll modify the code to match whichever strategy you prefer.

adolfo-rt replied:
I'm realizing just now that this is an API-breaking change. If we choose to modify the API, we should then be targetting the next ROS distro, J-turtle. The practice we've been following so far to reduce developer pain is to keep API stability over a ROS release, but allow to break ABI compatibility if needed.

API stability aside, I like the idea of exposing the whole SDF to `RobotHWSim` instances so they can do plugin-specific parsing. It's the resulting `initSim` signature that I don't like much, as it has pre-parsed info like the robot namespace, which comes also from the SDF. I'm also not very fond of it having 5 (with this PR, 6) parameters, it's too much for my taste.

Straying a bit from the scope of this PR (and also thinking of the next ROS release) maybe the `initSim` signature would benefit from an API rework. I'm thinking about having a lean signature like

``` c++
virtual bool initSim(gazebo::physics::ModelPtr parent, const sdf::ElementPtr sdf) = 0;
```

and then a bunch of simple, stateless helper functions to fill in the gap, like:

``` c++
std::string getRobotNamespace(const sdf::ElementPtr sdf);
std::string getRobotParam(const sdf::ElementPtr sdf);
const urdf::Model* getUrdf(const std::string& robot_param);
std::vector<transmission_interface::TransmissionInfo> parseTransmissions(const std::string& urdf_string);
ros::NodeHandle getModelNodeHandle(const std::string& robot_ns); // Too trivial to even expose?
```

What are your thoughts @jim-rothrock, @jbohren, @davetcoleman?.

adolfo-rt replied:
We could also parse the e-stop topic name from the ROS param server and not break the API, but that would fragment the configuration of a `RobotHWSim` plugin to multiple places (SDF, param server), which I think is undesirable.

jim-rothrock replied:
> What are your thoughts @jim-rothrock, @jbohren, @davetcoleman?

The new signature for initSim() sounds good to me. The current PR, though, can be implemented without modifying the API if I use strategy 1, described above. All the topic work is performed in gazebo_ros_control_plugin.cpp. All DefaultRobotHWSim would do is check its e_stop_active_ member variable, and set joint values accordingly. RobotHWSim::eStopActive()'s signature would change so that it takes a bool and returns void. The default implementation would be an empty function.

davetcoleman replied:
> If we choose to modify the API, we should then be targetting the next ROS distro, J-turtle

I was going to say the same thing. It is probably safest to put all new features there, in fact

+1 API rework, that is much cleaner.

adolfo-rt replied:
If strategy 1) does not break the API, then it could likely be merged into indigo. I opened #280 to not forget about the API braking proposal.

adolfo-rt commented on gazebo_ros_control/src/gazebo_ros_control_plugin.cpp:
```diff
@@ -206,7 +207,18 @@ void GazeboRosControlPlugin::Update()
+    bool reset_ctrlrs;
+    if (robot_hw_sim_->eStopActive())
+    {
+      e_stop_active_ = true;
```
Isn't this statement a no-op?. In the above if, `robot_hw_sim_->eStopActive()` is in fact checking the value of `e_stop_active_`

adolfo-rt replied:
OK, I got confused, ignore the above comment. There are `e_stop_active_` members in both `GazeboRosControlPlugin` and in `DefaultRobotHWSim`.

adolfo-rt commented:
Apologies for the delayed review. Comments inline.

davetcoleman commented:
Could you add a link to more info about SDF estop feature? I looked on sdformat.org just now and didn't find it

davetcoleman commented on gazebo_ros_control/src/default_robot_hw_sim.cpp:
```diff
@@ -335,11 +362,12 @@ class DefaultRobotHWSim : public gazebo_ros_control::RobotHWSim
         case VELOCITY_PID:
-          const double error = joint_velocity_command_[j] - joint_velocity_[j];
+          const double error = e_stop_active_ ? -joint_velocity_[j] :
+            joint_velocity_command_[j] - joint_velocity_[j];
```
Readability is tricky here, can we use a full `if` statement please?

jim-rothrock replied:
> Could you add a link to more info about SDF estop feature?

E-stop is a feature of ros_control, not SDF. As far as I know, it is documented only in @adolfo-rt's 2014 ROSCon presentation. Go to http://vimeo.com/107507546. Adolfo begins talking about ros_control's E-stop feature at 27:14. We have been discussing SDF because that's where the parameter that defines the E-stop control topic is specified.

jim-rothrock replied:
> Readability is tricky here, can we use a full if statement please?

Okay.

davetcoleman commented on gazebo_ros_control/src/default_robot_hw_sim.cpp:
```diff
@@ -500,10 +540,14 @@ class DefaultRobotHWSim : public gazebo_ros_control::RobotHWSim
   std::vector<double> joint_effort_;
   std::vector<double> joint_effort_command_;
-  std::vector<double> joint_position_command_;
+  std::vector<double> joint_position_command_, last_joint_position_command_;
```
just make a new line for this?

jim-rothrock replied:
> just make a new line for this?

Okay.

Commit:
Moved the E-stop callback from default_robot_hw_sim.cpp to
gazebo_ros_control_plugin.cpp. Removed the "sdf" parameter from
RobotHWSim::initSim() in order to avoid an API change.

Commit:
Removed model_nh_. It is no longer used.

jim-rothrock commented:
> If strategy 1) does not break the API, then it could likely be merged into indigo. I opened #280 to not
> forget about the API braking proposal.

Modified the code so that it does not break the API. Addressed other comments. Ready to be merged.

davetcoleman commented:
+1
Adolfo can you merge it?

adolfo-rt commented:
Merging.

Many thanks for adding the feature in an API compatible way, and for the very useful discussion on how we could implement this in a cleaner way when API breaks are allowed.

adolfo-rt commented:
Oops, Github can't automatically merge the PR. @jim-rothrock, is your branched synced with `indigo-devel`?

Commit:
Merged the E-stop code into gazebo_ros_control.

Commit:
Restored the formatting of registerJointLimits().

Commit:
Restored the formatting of registerJointLimits().

jim-rothrock commented:
> Oops, Github can't automatically merge the PR. @jim-rothrock, is your branched synced with indigo-devel?

I have modified the code so that it merges with the changes introduced in subsequent pull requests. Try merging it again.

adolfo-rt commented:
b88970a brings in a lot of code not written by you. Did you do the merge by hand?. Could you not just rebase your branch against `indigo-devel` so your changes are applied on top of it?. You could try this in a separate testing branch to make sure you don't break anything on your devel branch.

jim-rothrock commented:
> b88970a brings in a lot of code not written by you. Did you do the merge by hand?

I did "git fetch upstream indigo-devel", then merged default_robot_hw_sim.h (didn't exist when I submitted the PR) and default_robot_hw_sim.cpp (a class was moved from this file to default_robot_hw_sim.h, preventing automatic merging) by hand. I don't know what caused all those changes in b88970a.

> Could you not just rebase your branch against indigo-devel so your changes are applied on top of it?

Initially, I couldn't because my changes to default_robot_hw_sim.cpp had to be manually distributed between default_robot_hw_sim.h and default_robot_hw_sim.cpp. Also, I didn't know about rebase until you mentioned it in your message. Most of my SCM experience is with svn/cvs/rcs/sccs. I'm now going to re-fork the repo and reimplement my changes.

jim-rothrock commented:
I'm closing this pull request. I'll open a new one based on the changes I just applied to a new fork.

adolfo-rt commented:
I'm so very grateful for #283. Thanks @jim-rothrock!
